Teach rustc about the Xtensa VaListImpl

MabezDev opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r​? <reviewer name>
-->
<!-- homu-ignore:end -->

Following on from the target Xtensa target PRs (https://github.com/rust-lang/rust/pull/125141, https://github.com/rust-lang/rust/pull/126380), this PR teaches rustc about the structure of the VA list on the Xtensa arch, as well as adding the required lowering to be able to actually use it.

kelnos committed:
Teach rust core about Xtensa VaListImpl and add a custom lowering of vaarg for xtensa.

LLVM does not include an implementation of the va_arg instruction for
Xtensa. From what I understand, this is a conscious decision and
instead language frontends are encouraged to implement it themselves.
The rationale seems to be that loading values correctly requires
language and ABI-specific knowledge that LLVM lacks.

This is true of most architectures, and rustc already provides
implementation for a number of them. This commit extends the support to
include Xtensa.

See https://lists.llvm.org/pipermail/llvm-dev/2017-August/116337.html
for some discussion on the topic.

Unfortunately there does not seem to be a reference document for the
semantics of the va_list and va_arg on Xtensa. The most reliable source
is the GCC implementation, which this commit tries to follow. Clang also
provides its own compatible implementation.

This was tested for all the types that rustc allows in variadics.

Co-authored-by: Brian Tarricone <brian@tarricone.org>
Co-authored-by: Jonathan Bastien-Filiatrault <joe@x2a.org>
Co-authored-by: Paul Lietar <paul@lietar.net>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9872173857/job/27261742538) [(plain)](https://github.com/rust-lang/rust/commit/dc02d76ab0b51ec39368dc9b50f923c3307e2411/checks/27261742538/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#12 [ 5/11] RUN npm install es-check@6.1.1 eslint@8.6.0 -g
#12 4.638 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 4.697 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 4.719 npm WARN deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
#12 4.758 npm WARN deprecated @humanwhocodes/config-array@0.9.5: Use @eslint/config-array instead
#12 4.768 npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
#12 4.778 npm WARN deprecated @humanwhocodes/object-schema@1.2.1: Use @eslint/object-schema instead
#12 5.710 
#12 5.710 added 205 packages in 6s
#12 5.710 
#12 5.710 20 packages are looking for funding
---
#16 3.609 Building wheels for collected packages: reuse
#16 3.610   Building wheel for reuse (pyproject.toml): started
#16 3.934   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.935   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#16 3.935   Stored in directory: /tmp/pip-ephem-wheel-cache-nhgk2d_n/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#16 3.938 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#16 3.959   Attempting uninstall: setuptools
#16 3.960     Found existing installation: setuptools 59.6.0
#16 3.961     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
   Compiling std v0.0.0 (/checkout/library/std)
error: unexpected `cfg` condition value: `xtensa`
  --> library/core/src/ffi/va_list.rs:19:13
   |
19 |         not(target_arch = "xtensa"),
   |
   |
   = note: expected values for `target_arch` are: `aarch64`, `arm`, `arm64ec`, `avr`, `bpf`, `csky`, `hexagon`, `loongarch64`, `m68k`, `mips`, `mips32r6`, `mips64`, `mips64r6`, `msp430`, `nvptx64`, `powerpc`, `powerpc64`, `riscv32`, `riscv64`, `s390x`, `sparc`, `sparc64`, `wasm32`, `wasm64`, `x86`, and `x86_64`
   = note: `-D unexpected-cfgs` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(unexpected_cfgs)]`

error: unexpected `cfg` condition value: `xtensa`
error: unexpected `cfg` condition value: `xtensa`
  --> library/core/src/ffi/va_list.rs:42:13
   |
42 |         not(target_arch = "xtensa"),
   |
   |
   = note: expected values for `target_arch` are: `aarch64`, `arm`, `arm64ec`, `avr`, `bpf`, `csky`, `hexagon`, `loongarch64`, `m68k`, `mips`, `mips32r6`, `mips64`, `mips64r6`, `msp430`, `nvptx64`, `powerpc`, `powerpc64`, `riscv32`, `riscv64`, `s390x`, `sparc`, `sparc64`, `wasm32`, `wasm64`, `x86`, and `x86_64`

error: unexpected `cfg` condition value: `xtensa`
   --> library/core/src/ffi/va_list.rs:120:7
    |
    |
120 | #[cfg(target_arch = "xtensa")]
    |
    |
    = note: expected values for `target_arch` are: `aarch64`, `arm`, `arm64ec`, `avr`, `bpf`, `csky`, `hexagon`, `loongarch64`, `m68k`, `mips`, `mips32r6`, `mips64`, `mips64r6`, `msp430`, `nvptx64`, `powerpc`, `powerpc64`, `riscv32`, `riscv64`, `s390x`, `sparc`, `sparc64`, `wasm32`, `wasm64`, `x86`, and `x86_64`

error: unexpected `cfg` condition value: `xtensa`
   --> library/core/src/ffi/va_list.rs:179:13
    |
    |
179 |         not(target_arch = "xtensa"),
    |
    |
    = note: expected values for `target_arch` are: `aarch64`, `arm`, `arm64ec`, `avr`, `bpf`, `csky`, `hexagon`, `loongarch64`, `m68k`, `mips`, `mips32r6`, `mips64`, `mips64r6`, `msp430`, `nvptx64`, `powerpc`, `powerpc64`, `riscv32`, `riscv64`, `s390x`, `sparc`, `sparc64`, `wasm32`, `wasm64`, `x86`, and `x86_64`

error: unexpected `cfg` condition value: `xtensa`
   --> library/core/src/ffi/va_list.rs:200:9
    |
    |
200 |         target_arch = "xtensa",
    |         ^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: expected values for `target_arch` are: `aarch64`, `arm`, `arm64ec`, `avr`, `bpf`, `csky`, `hexagon`, `loongarch64`, `m68k`, `mips`, `mips32r6`, `mips64`, `mips64r6`, `msp430`, `nvptx64`, `powerpc`, `powerpc64`, `riscv32`, `riscv64`, `s390x`, `sparc`, `sparc64`, `wasm32`, `wasm64`, `x86`, and `x86_64`

error: unexpected `cfg` condition value: `xtensa`
   --> library/core/src/ffi/va_list.rs:146:17
    |
    |
146 |             not(target_arch = "xtensa"),
    |
    |
    = note: expected values for `target_arch` are: `aarch64`, `arm`, `arm64ec`, `avr`, `bpf`, `csky`, `hexagon`, `loongarch64`, `m68k`, `mips`, `mips32r6`, `mips64`, `mips64r6`, `msp430`, `nvptx64`, `powerpc`, `powerpc64`, `riscv32`, `riscv64`, `s390x`, `sparc`, `sparc64`, `wasm32`, `wasm64`, `x86`, and `x86_64`

error: unexpected `cfg` condition value: `xtensa`
   --> library/core/src/ffi/va_list.rs:161:13
    |
    |
161 |             target_arch = "xtensa",
    |             ^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: expected values for `target_arch` are: `aarch64`, `arm`, `arm64ec`, `avr`, `bpf`, `csky`, `hexagon`, `loongarch64`, `m68k`, `mips`, `mips32r6`, `mips64`, `mips64r6`, `msp430`, `nvptx64`, `powerpc`, `powerpc64`, `riscv32`, `riscv64`, `s390x`, `sparc`, `sparc64`, `wasm32`, `wasm64`, `x86`, and `x86_64`

error: could not compile `core` (lib) due to 7 previous errors
Build completed unsuccessfully in 0:00:25
  local time: Wed Jul 10 10:09:57 UTC 2024
```

</details>

oli-obk commented:
r? @davidtwco

tgross35 commented:
cc @workingjubilee who has been doing a lot of varargs work recently

workingjubilee commented on compiler/rustc_codegen_llvm/src/va_arg.rs:
```diff
@@ -271,6 +279,87 @@ fn emit_s390x_va_arg<'ll, 'tcx>(
+    list: OperandRef<'tcx, &'ll Value>,
+    target_ty: Ty<'tcx>,
+) -> &'ll Value {
+    // Implementation of va_arg for Xtensa. There doesn't seem to be an authoritative source for
+    // this, other than "what GCC does".
```
Fun.

workingjubilee reviewed (changes requested):
I haven't reviewed the codegen changes in detail as things will go much more smoothly if I have the bits of information I am asking for details on. If you don't have all this information, that's fine, and don't go off for a year digging for answers or anything. I just want as much of it that is on-hand and easily referenced.

tgross35 commented:
@rustbot author

MabezDev commented on compiler/rustc_codegen_llvm/src/va_arg.rs:
```diff
@@ -271,6 +279,87 @@ fn emit_s390x_va_arg<'ll, 'tcx>(
+    //
+    // The va_list type has three fields:
+    // struct __va_list_tag {
+    //   int32_t *va_stk; // Arguments passed on the stack
+    //   int32_t *va_reg; // Arguments passed in registers, saved to memory by the prologue.
```
Yes, structs can be passed via registers, structs can even be split across registers and stack space, _but_ primitive types will never be split. For instance, a `u128` either has to be all in registers, or all in stack never split across both.

> The reason I'm asking is because I want to know if this is one of those varargs ABIs where extern "C" fn (c_int, ...) where only 1 arg, a c_int, is passed, is ABI-equivalent to extern "C" fn (c_int, c_int) on entry into the actual function. Some ABIs deliberately design their varargs implementation to work like that, allowing you to basically deal with those two function pointers interchangeably. Others do not allow such because all varargs are handled in a distinct way so you only would match ABI details between varargs and other functions with only fixed args by accident.

From the Xtensa ISA:

![image](https://github.com/user-attachments/assets/f87fd0ec-0156-4e20-bd78-a97055c2e5e7)

So I believe in this case calling a normal function or vararg function is the same, but like you mentioned the function then has to unpack it's arguments in the vararg case.

MabezDev commented:
Gentle ping for a review whenever you have time some spare cycles :)

workingjubilee reviewed (changes requested):
At each step, I prefer to confirm the code implements the spec rather than read the code and try to discern what spec it implements. There's only so many ways to implement a spec, but there's an unbounded set of specs a given fragment could implement. So, this could use a few more comments.

Perhaps a less overloaded naming scheme would suffice? But when we're already checking in about 100 lines of code, we can afford a bit of extra padding to smooth things for the next reader.

alex-semenyuk commented:
@MabezDev
From wg-triage. Do you have any updates on this PR?

MabezDev commented on compiler/rustc_codegen_llvm/src/va_arg.rs:
```diff
@@ -271,6 +279,88 @@ fn emit_s390x_va_arg<'ll, 'tcx>(
+
+    // Update the offset in va_list, by adding the slot's size.
+    let offset_next = bx.add(offset, bx.const_i32(slot_size));
+    bx.store(offset_next, offset_ptr, bx.tcx().data_layout.pointer_align.abi);
```
I moved the store to the reg branch, there is a different once that applies for stack arguments.

MabezDev commented on compiler/rustc_codegen_llvm/src/va_arg.rs:
```diff
@@ -271,6 +279,88 @@ fn emit_s390x_va_arg<'ll, 'tcx>(
+    let offset_next = bx.add(offset, bx.const_i32(slot_size));
+    bx.store(offset_next, offset_ptr, bx.tcx().data_layout.pointer_align.abi);
+
+    // Figure out where to look for our value. We do that by checking the end of our slot (offset_next).
+    // If that is within the regsave area, then load from there. Otherwise load from the stack area.
+    let regsave_size = bx.const_i32(24);
+    let use_regsave = bx.icmp(IntPredicate::IntULE, offset_next, regsave_size);
+    bx.cond_br(use_regsave, from_regsave, from_stack);
```
I'm not sure of a better way to do this, `ndx` needs to be updated when we switch from reg args to stack args, so the check is definitely required, but we don't need to do it more than once because once we're using stack memory we won't ever touch the registers again. Do you have any suggestions to avoid doing this check more than once?

workingjubilee replied:
so, we have a model that, assuming all `va_arg` reads happened in immediate succession, could look like this, right?

```
va_arg_start:
    // code
    branch_CC   from_stack
    branch      from_regsave

from_regsave:
    // code
    branch_CC   from_stack
    branch      from_regsave

from_stack:
    // code
    branch      from_stack
```

hmm... I have a feeling that there is not a good way without complicating the implementation greatly, so I will not expect that here.

MabezDev commented:
Thank you for the review @workingjubilee (and sorry for the delay), I think I've addressed most of your initial concerns. I'm a little unsure on how to improve the stack ndx check, so any ideas would be appreciated.

workingjubilee commented on compiler/rustc_codegen_llvm/src/va_arg.rs:
```diff
@@ -270,6 +278,93 @@ fn emit_s390x_va_arg<'ll, 'tcx>(
+
+    // Figure out where to look for our value. We do that by checking the end of our slot (offset_next).
+    // If that is within the regsave area, then load from there. Otherwise load from the stack area.
+    let regsave_size = bx.const_i32(6);
```
Hmm. This changed from 24 to 6... did I miss something the first time?

We align-up unconditionally, now...

The offset still seems to have the same range of values...

I must have, if this change is correct?

MabezDev replied:
Apologies, that was a mistake. I misunderstood what the `const_i32` method was doing. I've reverted the change.

workingjubilee reviewed (changes requested):
Currently I still feel like I'm missing something or misreading something, and there is at least one clear omission.

MabezDev commented:
Thank you for the review! You were correct in your understanding of the LLVM codegen API, and I have added the relevant comments above the blocks in Rust code. This was a nice suggestion as it makes it a lot more readable.

workingjubilee commented on compiler/rustc_codegen_llvm/src/va_arg.rs:
```diff
@@ -270,6 +278,110 @@ fn emit_s390x_va_arg<'ll, 'tcx>(
+
+    bx.switch_to_block(from_stack);
+
+    // The first time we switch from regsave to stack we needs to adjust our offsets a bit.
+    // va_stk is set up such that the first stack argument is always at va_stk + 32.
+    // The corrected offset is written back into the va_list struct.
+
+    // let offset_corrected = cmp::max(offset, 32);
+    let needs_correction = bx.icmp(IntPredicate::IntULE, offset, regsave_size);
+    let offset_corrected = bx.select(needs_correction, bx.const_i32(32), offset);
```
If we actually want the `max` semantics, then this needs to also compare against `bx.const_i32(32)` and not `regsave_size`.

workingjubilee commented on compiler/rustc_codegen_llvm/src/va_arg.rs:
```diff
@@ -270,6 +278,110 @@ fn emit_s390x_va_arg<'ll, 'tcx>(
+    let needs_correction = bx.icmp(IntPredicate::IntULE, offset, regsave_size);
+    let offset_corrected = bx.select(needs_correction, bx.const_i32(32), offset);
+
+    // let offset_next_corrected = if offset <= 24 {
+    //     32 + slot_size
+    // } else {
+    //     offset + slot_size
+    // };
+    // va_ndx = offset_next_corrected;
+    let offset_next_corrected =
+        bx.select(needs_correction, bx.const_i32(32 + slot_size), offset_next);
```
The reason I pointed out this framing is that this makes it more obvious that this expression *should* be equivalent to just
```rust
let offset_next_corrected = offset_corrected + slot_size;
```

and thus should be
```suggestion
    // let offset_next_corrected = offset_corrected + slot_size;
    // va_ndx = offset_next_corrected;
    let offset_next_corrected = bx.add(offset_next, slot_size);
```

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12075504923/job/33675414851) [(plain)](https://github.com/rust-lang/rust/commit/aa5c9bce74bc49d385c7b7e8f5a2e85ecbd231d3/checks/33675414851/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 27.8s done
#16 DONE 41.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
   Compiling rustc_hir_typeck v0.0.0 (/checkout/compiler/rustc_hir_typeck)
error[E0308]: mismatched types
   --> compiler/rustc_codegen_llvm/src/va_arg.rs:354:53
    |
354 |     let offset_next_corrected = bx.add(offset_next, slot_size);
    |                                    ---              ^^^^^^^^^ expected `&llvm_::ffi::Value`, found `i32`
    |                                    arguments to this method are incorrect
    |
   ::: compiler/rustc_codegen_llvm/src/llvm/ffi.rs:685:5
    |
    |
685 |     pub type Value;
    |     -------------- the expected foreign type
    |
    = note: expected reference `&'ll llvm_::ffi::Value`
note: method defined here
   --> /checkout/compiler/rustc_codegen_ssa/src/traits/builder.rs:127:8
    |
    |
127 |     fn add(&mut self, lhs: Self::Value, rhs: Self::Value) -> Self::Value;

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_codegen_llvm` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

MabezDev commented:
Thanks for bearing with me, and sorry for the miscommunication. I believe I've addressed your comments properly now, let me know if there is anything else you'd like me to adjust.

workingjubilee commented on compiler/rustc_codegen_llvm/src/va_arg.rs:
```diff
@@ -270,6 +278,106 @@ fn emit_s390x_va_arg<'ll, 'tcx>(
+
+    // On big-endian, for values smaller than the slot size we'd have to align the read to the end
+    // of the slot rather than the start. While the ISA and GCC support big-endian, all the Xtensa
+    // targets supported by rustc are litte-endian so don't worry about it.
```
```suggestion
    // targets supported by rustc are little-endian so don't worry about it.
```

workingjubilee reviewed (approved):
No problem!

I believe this is correct.

workingjubilee commented:
r=me with the typo settled. :3

@bors delegate+

MabezDev commented:
@bors r+

MabezDev commented:
@bors r=workingjubilee

MabezDev commented:
I think I broke bors, let me try again :sweat_smile:.

@bors r=workingjubilee

tgross35 commented:
Sometimes things just get out of sync

@bors r=workingjubilee

bors commented:
:pushpin: Commit 059f6272c346654dfce7a68a6bb0c0aeb1ed756a has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 059f6272c346654dfce7a68a6bb0c0aeb1ed756a -->
<!-- homu: {"type":"Approved","sha":"059f6272c346654dfce7a68a6bb0c0aeb1ed756a","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
